x/ccv/provider/keeper/distribution.go

```diff
@@ -1,10 +1,141 @@
 package keeper
 
 import (
+	"errors"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/interchain-security/x/ccv/provider/types"
+	ccv "github.com/cosmos/interchain-security/x/ccv/types"
 )
 
 func (k Keeper) GetFeeCollectorAddressStr(ctx sdk.Context) string {
 	return k.accountKeeper.GetModuleAccount(
 		ctx, k.feeCollectorName).GetAddress().String()
 }
+
+func (k Keeper) GetProviderDistributionAddressStr(ctx sdk.Context) string {
+	return k.accountKeeper.GetModuleAccount(
+		ctx, types.DistributionAccount).GetAddress().String()
+}
+
+func (k Keeper) HandleProviderPoolWeights(ctx sdk.Context, weights ccv.ProviderPoolWeights) error {
+	if weights.GetTokens().IsZero() {
+		return errors.New("Tokens should not be zero ")
+	}
+	if len(weights.Addresses) == 0 {
+		return errors.New("Addresses should not be empty ")
+	}
+	if len(weights.Addresses) != len(weights.Weights) {
+		return errors.New("Addresses size does not equal to weights size ")
+	}
+	totalWeights := sdk.ZeroInt()
+	for _, weight := range weights.Weights {
+		totalWeights = totalWeights.Add(weight)
+	}
+	if !totalWeights.Equal(weights.TotalWeight) {
+		return errors.New("TotalWeight does not equal to the sum of weights")
+	}
+
+	k.AddPendingProviderPoolWeights(ctx, &weights)
+	return nil
+}
+
+func (k Keeper) Distribute(ctx sdk.Context) {
+	address := k.accountKeeper.GetModuleAccount(
+		ctx, types.DistributionAccount).GetAddress()
+	balances := k.bankKeeper.GetAllBalances(ctx, address)
+	if balances.IsZero() {
+		return
+	}
+
+	pending := k.GetPendingProviderPoolWeights(ctx)
+	breakIndex := -1
+	for i, ppw := range pending.GetPoolWeights() {
+		if balances.Len() == 0 || ppw.Tokens.IsAnyGT(balances) {
+			breakIndex = i
+			break
+		}
+		// transfer all tokens to distribution module address
+		if err := k.bankKeeper.SendCoinsFromModuleToModule(ctx, types.DistributionAccount, k.distributionName, ppw.GetTokens()); err != nil {
+			panic(err)
+		}
+		balances = balances.Sub(ppw.Tokens)
+		k.AllocateTokens(ctx, *ppw)
+	}
+
+	if breakIndex == 0 { // no pending ppw can be handled
+		return
+	}
+	if breakIndex == -1 { // all pending ppw can be handled
+		k.SetPendingProviderPoolWeights(ctx, ccv.PendingPPWs{})
+		return
+	}
+	pending.PoolWeights = pending.PoolWeights[breakIndex:]
+	k.SetPendingProviderPoolWeights(ctx, pending)
```

Comment:
maybe we can join breakIndex < len(pending)-1 cases, coz all of it need to 
k.SetPendingProviderPoolWeights(ctx, pending) (pending may sliced)

Reply:
we need `k.SetPendingProviderPoolWeights(ctx, pending)` only when` breakIndex != 0` .  and the breakIndex represents the `index`  where the pending. PoolWeights needs to be pruned

Reply:
in this case, if you got the breakIndex != -1, such as 1, and you set the "pending.PoolWeights[1:]" slice back, the next time you still get the "ppw" object refer to this breakIndex and it will be 0. The function Distribute will never going through, due to "if breakIndex == 0 {return}". This breaks shall be careful, it may have some better process way instead of the breakIndex

Reply:
1. if all the pending PPW are satisfied to be handled(balance tokens is gt tokens in PPW), then breakIndex is -1 as the initial value;
2. if breakIndex is 0, means the first pending PPW token is gt the account balance, thus the other pending PPWs are not satisfied either, so we return immediately;
3. Otherwise, some PPWs are satisfied while still has others PPW are not. so we keep the unsatisfied PPWs back to the store using 
```
pending.PoolWeights = pending.PoolWeights[breakIndex:]
k.SetPendingProviderPoolWeights(ctx, pending)
```

Reply:
If one "ppw" object match this "if balances.Len() == 0 || ppw.Tokens.IsAnyGT(balances) {" judgement, all distribute(including after blocks) will never go through since this ppw, all distribute stopped due to this "ppw" from this one and sofar. Is this what we want?